src/diffusers/models/resnet.py

```diff
@@ -606,6 +607,206 @@ def forward(self, hidden_states):
         return hidden_states * torch.tanh(torch.nn.functional.softplus(hidden_states))
 
 
+class InflatedConv3d(nn.Conv2d):
+    def forward(self, x):
+        video_length = x.shape[2]
+        # x = rearrange(x, "b c f h w -> (b f) c h w")
+        x = x.movedim((0, 1, 2, 3, 4), (0, 2, 1, 3, 4))
+        x = x.flatten(0, 1)
+        x = super().forward(x)
+        # x = rearrange(x, "(b f) c h w -> b c f h w", f=video_length)
+        x = x.reshape([-1, video_length, *x.shape[1:]])
+        x = x.movedim((0, 1, 2, 3, 4), (0, 2, 1, 3, 4))
+
+        return x
+
+
+class Upsample3D(nn.Module):
+    def __init__(self, channels, use_conv=False, use_conv_transpose=False, out_channels=None, name="conv"):
+        super().__init__()
+        self.channels = channels
+        self.out_channels = out_channels or channels
+        self.use_conv = use_conv
+        self.use_conv_transpose = use_conv_transpose
+        self.name = name
+
+        conv = None
+        if use_conv_transpose:
+            raise NotImplementedError
+        elif use_conv:
+            conv = InflatedConv3d(self.channels, self.out_channels, 3, padding=1)
+
+        if name == "conv":
+            self.conv = conv
+        else:
+            self.Conv2d_0 = conv
+
+    def forward(self, hidden_states, output_size=None):
+        assert hidden_states.shape[1] == self.channels
+
+        if self.use_conv_transpose:
+            raise NotImplementedError
+
+        # Cast to float32 to as 'upsample_nearest2d_out_frame' op does not support bfloat16
+        dtype = hidden_states.dtype
+        if dtype == torch.bfloat16:
+            hidden_states = hidden_states.to(torch.float32)
+
+        # upsample_nearest_nhwc fails with large batch sizes. see https://github.com/huggingface/diffusers/issues/984
+        if hidden_states.shape[0] >= 64:
+            hidden_states = hidden_states.contiguous()
+
+        # if `output_size` is passed we force the interpolation output
+        # size and do not make use of `scale_factor=2`
+        if output_size is None:
+            hidden_states = F.interpolate(hidden_states, scale_factor=[1.0, 2.0, 2.0], mode="nearest")
+        else:
+            hidden_states = F.interpolate(hidden_states, size=output_size, mode="nearest")
+
+        # If the input is bfloat16, we cast back to bfloat16
+        if dtype == torch.bfloat16:
+            hidden_states = hidden_states.to(dtype)
+
+        if self.use_conv:
+            if self.name == "conv":
```

Comment:
Yes, it's not doing anything because in the init function, both self.conv and self.Conv2d_0 are initialized the same. I'll get rid of it.

Reply:
Actually, this is because of a comment I found in the `ResnetBlock2D`:

# TODO(Suraj, Patrick) - clean up after weight dicts are correctly renamed
if name == "conv":
       self.conv = conv
else:
       self.Conv2d_0 = conv